### ISSUE TITLE
have problems with back-propagation

### ISSUE THREAD

**Issue by XiangMochu** [anucvml/ddn#2]
Hi there, I really appreciate your work and I see big potentials in it.

I'm trying to embed a DDN layer into my work, and I'm having trouble with it, which has been bothering me for weeks. It seems that my implementation of a DDN layer did not properly backpropagate gradients.

In detail, I want to use DDN to perform a least square minimization, say, $y=argmin_{u} \| uF-x \|_2$, while **allowing the gradient to backpropagate through the layer**. If I use a DDN layer to solve this problem, it seems that the network does not converge well; but if I compute $y$ with a closed-form solution, there is no such problem. The problem still exists even if I initialize the output using the closed-form solution in the DDN layer. The layer performs well in inference, so I assume that the accuracy is not a problem, but the back-propagating is, the DDN layer may block the gradient to be propagated to the network that's in front of it.

I'm not sure if I implemented it in the right way. When I implement the `solve` method, do I have to detach all the input variables? When I call the `solve` method, do I have to put it in the `torch.no_grad()`?  And, do I have to manually add `y.requires_grad_()` after `y` is solved? I have tried it with and without the above, but it did not seem to work properly, I think I must have missed something.

Looking forward to your reply.

**Comment by XiangMochu** [comment]
I just realized that I have to use the `DeclarativeLayer` instead of `AbstractDeclarativeNode` as the PyTorch layer, and the DDN seems to work just fine. It's a little bit confusing since the 'node' can also backpropagate gradients, but not in the desired way😰. Maybe add some hints to the notebook so that readers can realize this before wasting time on debugging.
